crates/order-validation/src/banned.rs

```diff
@@ -0,0 +1,91 @@
+use {
+    contracts::ChainalysisOracle,
+    ethcontract::{errors::MethodError, H160},
+    ethrpc::Web3,
+    std::{collections::HashSet, sync::RwLock, time::Duration},
+    ttl_cache::TtlCache,
+};
+
+/// A list of banned users and an optional registry that can be checked onchain.
+pub struct Users {
+    list: HashSet<H160>,
+    onchain: Option<Onchain>,
+}
+
+struct Onchain {
+    contract: ChainalysisOracle,
+    cache: RwLock<TtlCache<H160, bool>>,
+}
+
+const TTL: Duration = Duration::from_secs(60 * 60); // 1 hour
+
+impl Users {
+    /// Creates a new `Users` instance that checks the hardcoded list and uses
+    /// the given `web3` instance to determine whether an onchain registry of
+    /// banned addresses is available.
+    pub async fn new(web3: &Web3, banned_users: Vec<H160>) -> Self {
+        Self {
+            list: HashSet::from_iter(banned_users),
+            onchain: ChainalysisOracle::deployed(web3)
+                .await
+                .map(|contract| Onchain {
+                    contract,
+                    cache: RwLock::new(TtlCache::new(usize::MAX)),
+                })
+                .ok(),
+        }
+    }
+
+    /// Creates a new `Users` instance that passes all addresses.
+    pub fn none() -> Self {
+        Self {
+            list: HashSet::new(),
+            onchain: None,
+        }
+    }
+
+    /// Creates a new `Users` instance that passes all addresses except for the
+    /// ones in `list`.
+    pub fn from_set(list: HashSet<H160>) -> Self {
+        Self {
+            list,
+            onchain: None,
+        }
+    }
+
+    /// Returns `true` if the given address is banned or an error if the result
+    /// cannot be determined.
+    pub async fn is_banned(&self, address: H160) -> Result<bool, Error> {
+        if self.list.contains(&address) {
+            return Ok(true);
+        }
+
+        if let Some(onchain) = &self.onchain {
+            if let Some(result) = onchain.cache.read().expect("unpoisoned").get(&address) {
+                return Ok(*result);
+            }
+
+            let result = onchain.fetch(address).await?;
+            onchain
+                .cache
+                .write()
```

Comment:
There will be a lot of contention in the autopilot whenever we restart the system or the hour is over.
Since auction creation already happens in the background this might not be a huge deal, though.
Alternatively you could consider passing a slice of addresses and return a set of banned addresses. That way there will be less contention on the lock. Maybe it makes sense to simply measure the time we spend filtering first.

Reply:
I implemented a version of this (see latest PR). I find the code the code less straight forward to read (ie. scoping of lock variable being important). 

The other difference is that we are now swallowing the Error (instead of returning it). The call-sites do look nicer now though.

I have no strong preference. I don't think 1k lock holdings are going to be the bottleneck but you are right in principle.

Reply:
I agree that this approach tends to make the code harder to read and debug.
I'm totally fine with measuring first (e.g. in the autopilot filtering) to see if it's actually an issue or just premature optimization.